tasks/tasks.go

```diff
@@ -0,0 +1,577 @@
+	var pollChan <-chan struct{} = taskChannels.Poll
+	var stopChan <-chan struct{} = taskChannels.Stop
+
+	// the task deletion period is specified in seconds
+	deleteAfter := time.Duration(config.Service.DeleteAfter) * time.Second
+
+	// start scurrying around
+	for {
+		select {
+		case newTask := <-taskChan: // Add() called
+			newTask.Id = uuid.New()
+			tasks[newTask.Id] = newTask
+			taskIdChan <- newTask.Id
+			slog.Info(fmt.Sprintf("Added new transfer task %s.", newTask.Id.String()))
+		case taskId := <-taskIdChan: // Status() called
+			if task, found := tasks[taskId]; found {
+				statusChan <- task.Status
+			} else {
+				err := fmt.Errorf("Task %s not found!", taskId.String())
+				errorChan <- err
+			}
+		case <-pollChan: // time to move things along
+			for taskId, task := range tasks {
+				oldStatus := task.Status
+				err := task.Update()
+				if err != nil {
+					// we log task update errors but do not propagate them
+					slog.Error(err.Error())
+				}
+				if task.Status.Code != oldStatus.Code {
+					switch task.Status.Code {
+					case TransferStatusStaging:
+						slog.Info(fmt.Sprintf("Staging files for task %s.", task.Id.String()))
+					case TransferStatusActive:
+						slog.Info(fmt.Sprintf("Beginning transfer for task %s.", task.Id.String()))
+					case TransferStatusInactive:
+						slog.Info(fmt.Sprintf("Suspended transfer for task %s.", task.Id.String()))
+					case TransferStatusFinalizing:
+						slog.Info(fmt.Sprintf("Finalizing transfer for task %s.", task.Id.String()))
+					case TransferStatusSucceeded:
+						slog.Info(fmt.Sprintf("Transfer task %s completed successfully.", task.Id.String()))
+					case TransferStatusFailed:
+						slog.Info(fmt.Sprintf("Transfer task %s failed.", task.Id.String()))
+					}
+				}
+
+				// if the task completed a long enough time go, delete its entry
+				if task.Age() > deleteAfter {
+					slog.Debug(fmt.Sprintf("Purging task %s.", task.Id.String()))
+					delete(tasks, taskId)
+				} else { // update its entry
+					tasks[taskId] = task
+				}
+			}
+		case <-stopChan: // Stop() called
+			err := saveTasks(tasks, dataStore) // don't forget to save our state!
+			errorChan <- err
+			break
+		}
+	}
+}
+
+// this function sends a regular pulse on its poll channel until the global
+// variable running is found to be false
+func heartbeat(pollInterval time.Duration, pollChan chan<- struct{}) {
+	for {
+		time.Sleep(pollInterval)
+		pollChan <- struct{}{}
+		if !running {
+			break
+		}
+	}
+}
+
+// this function checks for the existence of the data directory and whether it
+// is readable/writeable, returning a non-nil error if any of these conditions
+// are not met
+func validateDataDirectory(dir string) error {
+	if dir == "" {
+		return fmt.Errorf("no data directory was specified!")
+	}
+	info, err := os.Stat(dir)
+	if err != nil {
+		return err
+	}
+	if !info.IsDir() {
+		return fmt.Errorf("Given data directory is not a directory!")
```

Comment:
Do you want to include the name of the data directory in the error message?

Reply:
I changed these to return `os.PathError`, which is suited to filesystem shenanigans.